pkg/controller/garbagecollector/rate_limiter_helper.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+// RegisteredRateLimiter records the registered RateLimters to avoid
+// duplication.
+type RegisteredRateLimiter struct {
+	rateLimiters map[unversioned.GroupVersion]struct{}
+	lock         sync.Mutex
+}
+
+// NewRegisteredRateLimiter returns a new RegisteredRateLimiater.
+func NewRegisteredRateLimiter() *RegisteredRateLimiter {
+	return &RegisteredRateLimiter{
+		rateLimiters: make(map[unversioned.GroupVersion]struct{}),
+		lock:         sync.Mutex{},
+	}
+}
+
+func (r *RegisteredRateLimiter) register(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
```

Comment:
probably registerIfNotPresent (or sth along those lines) would be a better name. I hope it won't degrade performance too badly...

I don't know if GC is multithreaded, but it there can be a lot of workers I'd probably change normal Lock to RWLock, do the check under RLock, then release RLock and take WLock, check if it's not present again to make sure nothing added a client in the meantime and register new metric. My hope would be that this would reduce contention on the lock.